sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/Crawler.scala

```diff
@@ -130,55 +135,101 @@ class Crawler extends CliTool {
         topN = topN, groupBy = groupBy)
       val fetchedRdd = rdd.map(r => (r.getGroup, r))
         .groupByKey()
         .flatMap({ case (grp, rs) => new FairFetcher(job, rs.iterator, localFetchDelay,
-          FetchFunction, ParseFunction, OutLinkFilterFunction) })
+          FetchFunction, ParseFunction, OutLinkFilterFunction)
+        })
         .persist()
 
       if (kafkaEnable) {
         storeContentKafka(kafkaListeners, kafkaTopic.format(jobId), fetchedRdd)
       }
 
-      //Step: Update status of fetched resources
+      //Step: Update status of fetched resources.
       val statusUpdateRdd: RDD[SolrInputDocument] = fetchedRdd.map(d => StatusUpdateSolrTransformer(d))
       val statusUpdateFunc = new SolrStatusUpdate(job)
       sc.runJob(statusUpdateRdd, statusUpdateFunc)
 
-      //Step: Filter Outlinks and Upsert new URLs into CrawlDb
-      val outlinksRdd = OutLinkUpsert(job, fetchedRdd)
-      val upsertFunc = new SolrUpsert(job)
-      sc.runJob(outlinksRdd, upsertFunc)
-
+      val scoredRdd = score(fetchedRdd)
       //Step: Store these to nutch segments
       val outputPath = this.outputPath + "/" + taskId
 
-      storeContent(outputPath, fetchedRdd)
+      storeContent(outputPath, scoredRdd)
 
       LOG.info("Committing crawldb..")
       solrc.commitCrawlDb()
     }
+
     solrc.close()
     //PluginService.shutdown(job)
     LOG.info("Shutting down Spark CTX..")
     sc.stop()
   }
+
+  def executePipeline(rdd: RDD[Resource], taskId: String, solrc: SolrProxy, localFetchDelay: Long): Unit = {
+    val fetchedRdd = rdd.map(r => (r.getGroup, r))
+      .groupByKey()
+      .flatMap({ case (grp, rs) => new FairFetcher(job, rs.iterator, localFetchDelay,
+        FetchFunction, ParseFunction, OutLinkFilterFunction)
+      })
+      .persist()
+
+    if (kafkaEnable) {
+      storeContentKafka(kafkaListeners, kafkaTopic.format(jobId), fetchedRdd)
+    }
+
+    //Step: Update status of fetched resources.
+    val statusUpdateRdd: RDD[SolrInputDocument] = fetchedRdd.map(d => StatusUpdateSolrTransformer(d))
+    val statusUpdateFunc = new SolrStatusUpdate(job)
+    sc.runJob(statusUpdateRdd, statusUpdateFunc)
+
+    val scoredRdd = score(fetchedRdd)
+    //Step: Store these to nutch segments
+    val outputPath = this.outputPath + "/" + taskId
+
+    storeContent(outputPath, scoredRdd)
+
+    LOG.info("Committing crawldb..")
+    solrc.commitCrawlDb()
+  }
+
+  def score(fetchedRdd: RDD[CrawlData]): RDD[CrawlData] = {
+    val job = this.job
+
+    val scoredRdd = fetchedRdd.map(d => ScoreFunction(job, d))
+
+    val scoreUpdateRdd: RDD[SolrInputDocument] = scoredRdd.map(d => ScoreUpdateSolrTransformer(d))
+    val scoreUpdateFunc = new SolrStatusUpdate(job)
+    sc.runJob(scoreUpdateRdd, scoreUpdateFunc)
+
+    //TODO (was OutlinkUpsert)
+    val outlinksRdd = scoredRdd.flatMap({ data => for (u <- data.parsedData.outlinks) yield (u, data.fetchedData.getResource) }) //expand the set
```

Comment:
why did we have to do this. There is already an `OutlinkUpsert`, which is now unused and duplicate code.
If it was me, I would have just let it the way it was.
i.e. not to delete the line  `val outlinksRdd = OutLinkUpsert(job, fetchedRdd)`